keps/sig-node/0014-runtime-class.md

```diff
@@ -0,0 +1,332 @@
+---
+kep-number: 14 FIXME(13)
+title: Runtime Class
+authors:
+  - "@tallclair"
+owning-sig: sig-node
+participating-sigs:
+  - sig-architecture
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2018-06-19
+status: provisional
+---
+
+# Runtime Class
+
+## Table of Contents
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non\-Goals](#non-goals)
+  * [User Stories](#user-stories)
+* [Proposal](#proposal)
+  * [API](#api)
+    * [Runtime Handler](#runtime-handler)
+  * [Versioning, Updates, and Rollouts](#versioning-updates-and-rollouts)
+  * [Implementation Details](#implementation-details)
+  * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Appendix](#appendix)
+  * [Examples of runtime variation](#examples-of-runtime-variation)
+
+## Summary
+
+`RuntimeClass` is a new cluster-scoped resource that surfaces container runtime properties to the
+control plane. RuntimeClasses are assigned to pods through a `runtimeClass` field on the
+`PodSpec`. This provides a new mechanism for supporting multiple runtimes in a cluster and/or node.
+
+## Motivation
+
+RuntimeClasses address two distinct problems: surfacing support and configuration of optional
+features, and supporting multiple runtimes.
+
+Today's runtimes have varying support for features that are expressible through the PodSpec (see
+[examples of runtime variation](#examples-of-runtime-variation) below). In some cases such as
+SELinux, features are silently ignored when they aren't supported. This is problematic from a
+security perspective since it makes it difficult to audit actual runtime options. Other features
+such as AppArmor are checked for support, but on failure the pod is just held in a "Blocked" state
+to avoid the churn of recreating the rejected pod. Still other features maybe be applied in slightly
+different configurations, as is the case with capabilities today (the default set is not
+well-defined), and `privileged` containers. By raising these properties to the control plane errors
+can be surfaced much earlier (admission), and configuration can be made explicit.
+
+The second problem that RuntimeClass addresses is supporting multiple runtimes in a cluster, or even
+on the same node. [Sandboxes][] are the primary motivator for this right now, with both Kata
+containers and gVisor looking to integrate with Kubernetes. Other runtime models such as Windows
+containers or even remote runtimes will also require support in the future. RuntimeClass provides a
```

Comment:
Windows may not be the best example unless we plan not to surface the information on the pod or node.

Reply:
I'm not sure I understand this. Do you mean we might want a top-level `OS` field on the podspec?

Reply:
its inevitable to me that we will need to declare the target os of a pod spec to simplify management of associated features.  defaulting must differ based on target platform and right now there is nothing but an opaque label in many cases to know if this is windows/linux.  to me, its inevitable that we will need to know the target platform on the pod spec itself similar to https://github.com/opencontainers/runtime-spec/blob/master/config.md#platform-specific-configuration

Reply:
What I was trying to say is that "Windows" is not just container runtime feature. It can require Windows node (i.e., host OS), and as @derekwaynecarr pointed out, specific Windows configurations in the pod spec. Given that, there are more obvious signals that we can/should use for pod scheduling. Using RuntimeClass to capture node-level property may be out of scope?

Reply:
This sounds like you would have to some sort of validation in where the runtime specified in the RuntimeClass would have to comply with the NodeSpec and/or PodSpec. But this may contradict with the scope creep like @derekwaynecarr was mentioning.